api/core/v1alpha2/virtual_machine_snapshot.go

```diff
@@ -0,0 +1,131 @@
+You may obtain a copy of the License at
+
+     http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// +kubebuilder:object:generate=true
+// +groupName=virtualization.deckhouse.io
+package v1alpha2
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+const (
+	VirtualMachineSnapshotKind     = "VirtualMachineSnapshot"
+	VirtualMachineSnapshotResource = "virtualmachinesnapshots"
+)
+
+// VirtualMachineSnapshot provides a resource for creating snapshots of virtual machines.
+//
+// +kubebuilder:object:root=true
+// +kubebuilder:metadata:labels={heritage=deckhouse,module=virtualization}
+// +kubebuilder:subresource:status
+// +kubebuilder:resource:categories=virtualization,scope=Namespaced,shortName={vmsnapshot,vmsnapshots},singular=virtualmachinesnapshot
+// +kubebuilder:printcolumn:name="Phase",type="string",JSONPath=".status.phase",description="VirtualMachineSnapshot phase."
+// +kubebuilder:printcolumn:name="Consistent",type="boolean",JSONPath=".status.consistent",description="VirtualMachineSnapshot consistency."
+// +kubebuilder:printcolumn:name="Age",type="date",JSONPath=".metadata.creationTimestamp",description="VirtualMachineSnapshot age."
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+type VirtualMachineSnapshot struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec   VirtualMachineSnapshotSpec   `json:"spec"`
+	Status VirtualMachineSnapshotStatus `json:"status,omitempty"`
+}
+
+// VirtualMachineSnapshotList contains a list of `VirtualMachineSnapshot`
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+type VirtualMachineSnapshotList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata"`
+	Items           []VirtualMachineSnapshot `json:"items"`
+}
+
+type VirtualMachineSnapshotSpec struct {
+	// The name of virtual machine to take snapshot.
+	//
+	// +required
+	VirtualMachineName string `json:"virtualMachineName"`
+	// Create a snapshot of a virtual machine only if it is possible to freeze the machine through the agent.
+	//
+	// If value is true, the snapshot of the virtual machine will be taken only in the following scenarios:
+	// - the virtual machine is powered off.
+	// - the virtual machine with an agent, and the freeze operation was successful.
+	//
+	// +kubebuilder:default:=true
+	RequiredConsistency bool `json:"requiredConsistency"`
+	// +kubebuilder:default:="Always"
+	KeepIPAddress KeepIPAddress `json:"keepIPAddress"`
+	// +optional
+	VolumeSnapshotClassNames []VolumeSnapshotClassName `json:"volumeSnapshotClassNames"`
```

Comment:
```
volumeSnapshotClassNames:
  - volumeSnapshotClassName: foo
    storageClassName: sc-foo
```
Looks weird. I think we can omit Names from parent field and call it `volumeSnapshotClasses`

```
```
volumeSnapshotClasses:
  - volumeSnapshotClassName: foo
    storageClassName: sc-foo
```